README.md
# Final-Project-CSPS
Scientific Python Interactive Data Acoustic Modeling (SPIDAM) Project

Our project is a Python application with the purpose of identifying the frequency ranges that have the longest reverb times (RT60), in order to implement the proper acoustic treatment. Created to improve voice intelligibility.

*read requirement.txt to find what to pip install.

Basic Usage: 
1)Upload audio file(.wav or .mp3).
2)Select the graph you want to display.


CHANGELOG.md
## [1.0] - 2023-12-10

- Combined plots.py and clean.py are combined to make model.py
- Final audio test file (our clap) was added

## [0.4] - 2023-12-09

- Wav forms are now analzed through plots.py
- Low, mid, and high frequencies can be displayed individually
- Low, mid, and high frequencies can be displayed all together
- Waveform can be displayed
- Spectrogram can be displayed
- Buttons now are able to control images displayed
- Title label has been added for images

## [0.3] - 2023-12-08

- GUI is able to import mp3 and wav files
- Cleans files of metadata
- Converts mp3s to wav
- returns cleaned and converted wav with file reference

## [0.2] - 2023-12-07

- GUI is now able to host pictures and graphs on frame
- Currently displays an image of BIG LION but will not later on
- Is able to take reference to imported files
- Is able to resize window based on size of labels, buttons, and images

## [0.1] - 2023-12-05

- Created Basic GUI


clean.py
from scipy.io import wavfile
import numpy as np
from pydub import AudioSegment
import os
import matplotlib as plt


def convert_to_wav(input_file, output_file):
    audio = AudioSegment.from_file(input_file)

    # Export the audio as a WAV file
    audio.export(output_file, format='wav')
    print(f"Conversion successful. Saved as {output_file}")

    return output_file

def clean_audio_data(file_path):


    def convert_to_mono(input_file, output_file):
        audio = AudioSegment.from_file(file_path)

        # Set channels to 1
        mono_audio = audio.set_channels(1)

        # Export the audio as a WAV file
        mono_audio.export(output_file, format='wav')
        print(f"Conversion successful. Saved as {output_file}")

        return output_file

    # Convert to WAV file if not already WAV
    if not file_path.lower().endswith('.wav'):
        file_path = convert_to_wav(file_path, 'converted_audio.wav')

    # Extract only samplerate and data
    samplerate, data = wavfile.read(file_path)

    # Check if channel format is more than 1, convert to mono
    if data.ndim > 1:
        file_path = convert_to_mono(file_path, 'mono_audio.wav')
        # Update with new sample rate and data values for mono
        samplerate, data = wavfile.read(file_path)

    # Handles missing values
    data = np.nan_to_num(data)

    newDirectory = os.path.dirname(os.path.abspath(__file__))
    # Create file path for new data
    cleaned_file_path = newDirectory + '\\cleaned_audio.wav'

    # Rewriting file ensures no meta data
    wavfile.write(cleaned_file_path, samplerate, data)
    print(f"Cleaning successful. Saved as {cleaned_file_path}")

    return cleaned_file_path

model.py
from scipy.io import wavfile
import numpy as np
from pydub import AudioSegment
import os
import numpy as np
import matplotlib.pyplot as plt
import io
from PIL import Image, ImageTk

class Model:
    def clean_audio_data(self, file_path):
        def convert_to_wav(input_file, output_file):
            audio = AudioSegment.from_file(input_file)

            # Export the audio as a WAV file
            audio.export(output_file, format='wav')
            print(f"Conversion successful. Saved as {output_file}")

            return output_file

        def convert_to_mono(input_file, output_file):
            audio = AudioSegment.from_file(input_file)

            # Set channels to 1
            mono_audio = audio.set_channels(1)

            # Export the audio as a WAV file
            mono_audio.export(output_file, format='wav')
            print(f"Conversion successful. Saved as {output_file}")

            return output_file

        # Convert to WAV file if not already WAV
        if not file_path.lower().endswith('.wav'):
            file_path = convert_to_wav(file_path, 'converted_audio.wav')

        # Extract only samplerate and data
        samplerate, data = wavfile.read(file_path)

        # Check if channel format is more than 1, convert to mono
        if data.ndim > 1:
            file_path = convert_to_mono(file_path, 'mono_audio.wav')
            # Update with new sample rate and data values for mono
            samplerate, data = wavfile.read(file_path)

        # Handles missing values
        data = np.nan_to_num(data)

        newDirectory = os.path.dirname(os.path.abspath(__file__))
        # Create file path for new data
        cleaned_file_path = newDirectory + '\\cleaned_audio.wav'

        # Rewriting file ensures no meta data
        wavfile.write(cleaned_file_path, samplerate, data)
        print(f"Cleaning successful. Saved as {cleaned_file_path}")

        return cleaned_file_path

    def fig2img(self, fig):
        buf = io.BytesIO()
        fig.savefig(buf)
        buf.seek(0)
        img = Image.open(buf)
        return img

    def createWavForm(self, file_path):
        plt.rcParams["figure.figsize"] = [7.50, 3.50]
        plt.rcParams["figure.autolayout"] = True
        plt.figure()

        # Read the .wav file
        sample_rate, data = wavfile.read(file_path)

        # Duration of the file
        duration = len(data) / sample_rate

        # Create time array for x-axis
        time = np.linspace(0., duration, len(data))

        # Plot the waveform
        plt.plot(time, data, lw=0.5)
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')

        fig = plt.gcf()
        plt.close('all')
        return ImageTk.PhotoImage(self.fig2img(fig)), duration

    def plotting(self, filepath, freq_target_range, printSpec, printAll):
        plt.rcParams["figure.figsize"] = [7.50, 3.50]
        plt.rcParams["figure.autolayout"] = True
        plt.figure()

        sample_rate, data = wavfile.read(filepath)
        spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
        plt.clf()

        duration = len(data) / sample_rate

        def debugg(fstring):
            print(fstring)

        def find_target_frequency(freqs):
            for x in freqs:
                if x > freq_target_range:
                    break
            return x

        def frequency_check():
            # debugg(f'freqs {freqs[:10]}]')
            target_frequency = find_target_frequency(freqs)
            # debugg(f'target_frequency {target_frequency}')
            index_of_frequency = np.where(freqs == target_frequency)[0][0]
            # debugg(f'index_of_frequency {index_of_frequency}')

            data_for_frequency = spectrum[index_of_frequency]
            # debugg(f'data_for_frequency {data_for_frequency[:10]}')

            data_in_db_fun = 10 * np.log10(data_for_frequency)
            return data_in_db_fun

        plt.figure()
        if (printAll == True):
            for i in range(3):
                freq_target_range = 2800 * (i ** 2) - (1900 * i) + 100
                data_in_db = frequency_check()
                match i:
                    case 0:
                        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#de3163', label='Low')
                    case 1:
                        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#00aaff', label='Med')
                    case 2:
                        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#800080', label='High')
            plt.legend()
        else:
            data_in_db = frequency_check()
            match freq_target_range:
                case 100:
                    plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#de3163')
                case 1000:
                    plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#00aaff')
                case 7500:
                    plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#800080')
                case _:
                    plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')

        plt.xlabel('Time (s)')
        plt.ylabel('Power (dB)')

        index_of_max = np.argmax(data_in_db)

        value_of_max = data_in_db[index_of_max]

        sliced_array = data_in_db[index_of_max:]

        value_of_max_less_5 = value_of_max - 5

        def find_nearest_value(array, value):
            array = np.asarray(array)
            idx = (np.abs(array - value)).argmin()
            return array[idx]

        value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)

        index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)

        value_of_max_less_25 = value_of_max - 25

        value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)

        index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

        rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

        rt60 = 3 * rt20

        # plt.grid()
        # plt.show()

        if (printSpec == True):
            plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
        elif (printAll == False):
            plt.plot(t[index_of_max], data_in_db[index_of_max], 'go')
            plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')
            plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')

        fig = plt.gcf()
        plt.close('all')
        return ImageTk.PhotoImage(self.fig2img(fig)), round(abs(rt60), 2), round(value_of_max, 2)



plots.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile
import io
import os
from PIL import Image, ImageTk

def fig2img(fig):
  buf = io.BytesIO()
  fig.savefig(buf)
  buf.seek(0)
  img = Image.open(buf)
  return img

def createWavForm(file_path):
  plt.rcParams["figure.figsize"] = [7.50, 3.50]
  plt.rcParams["figure.autolayout"] = True
  plt.figure()

  # Read the .wav file
  sample_rate, data = wavfile.read(file_path)

  # Duration of the file
  duration = len(data) / sample_rate

  # Create time array for x-axis
  time = np.linspace(0., duration, len(data))

  # Plot the waveform
  plt.plot(time, data, lw=0.5)
  plt.xlabel('Time (s)')
  plt.ylabel('Amplitude')

  fig = plt.gcf()
  plt.close('all')
  return ImageTk.PhotoImage(fig2img(fig)), duration

def plotting(filepath, freq_target_range, printSpec, printAll):
  plt.rcParams["figure.figsize"] = [7.50, 3.50]
  plt.rcParams["figure.autolayout"] = True
  plt.figure()

  sample_rate, data = wavfile.read(filepath)
  spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
  plt.clf()

  duration = len(data) / sample_rate

  def debugg(fstring):
    print(fstring)

  def find_target_frequency(freqs):
    for x in freqs:
      if x > freq_target_range:
        break
    return x

  def frequency_check():
    # debugg(f'freqs {freqs[:10]}]')
    target_frequency = find_target_frequency(freqs)
    # debugg(f'target_frequency {target_frequency}')
    index_of_frequency = np.where(freqs == target_frequency)[0][0]
    # debugg(f'index_of_frequency {index_of_frequency}')

    data_for_frequency = spectrum[index_of_frequency]
    # debugg(f'data_for_frequency {data_for_frequency[:10]}')

    data_in_db_fun = 10 * np.log10(data_for_frequency)
    return data_in_db_fun

  plt.figure()
  if (printAll == True):
    for i in range(3):
      freq_target_range = 2800 * (i ** 2) - (1900 * i) + 100
      data_in_db = frequency_check()
      match i:
        case 0:
          plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#de3163', label='Low')
        case 1:
          plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#00aaff', label='Med')
        case 2:
          plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#800080', label='High')
    plt.legend()
  else:
    data_in_db = frequency_check()
    match freq_target_range:
      case 100:
        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#de3163')
      case 1000:
        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#00aaff')
      case 7500:
        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#800080')
      case _:
        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')

  plt.xlabel('Time (s)')
  plt.ylabel('Power (dB)')

  index_of_max = np.argmax(data_in_db)

  value_of_max = data_in_db[index_of_max]

  sliced_array = data_in_db[index_of_max:]

  value_of_max_less_5 = value_of_max - 5

  def find_nearest_value(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return array[idx]

  value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)

  index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)

  value_of_max_less_25 = value_of_max - 25

  value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)

  index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

  rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

  rt60 = 3 * rt20

  #plt.grid()
  #plt.show()

  if (printSpec == True):
    plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
  elif (printAll == False):
    plt.plot(t[index_of_max], data_in_db[index_of_max], 'go')
    plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')
    plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')

  fig = plt.gcf()
  plt.close('all')
  return ImageTk.PhotoImage(fig2img(fig)), round(abs(rt60), 2), round(value_of_max, 2)

Reverb_Plot.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile
from scipy.signal import spectrogram

# Load the .wav file
sample_rate, data = wavfile.read("New_Recording_68.wav")

# Compute the spectrogram using FFT
frequencies, times, spectrum = spectrogram(data, fs=sample_rate, nperseg=1024)

# Select a frequency
def find_target_frequency(freqs, target):
    for x in freqs:
        if x > target:
            break
    return x

def frequency_check(freq_range):
    # Identify a frequency to check
    global target_frequency
    target_frequency = find_target_frequency(frequencies, freq_range)
    index_of_frequency = np.where(frequencies == target_frequency)[0][0]

    # Find sound data for a particular frequency
    data_for_frequency = spectrum[index_of_frequency]

    # Change a digital signal to values in decibels
    data_in_db_fun = 10 * np.log10(data_for_frequency)

    return data_in_db_fun

low = 100
mid = 1000
high = 10000

# Call the function to get the data in decibels
data_in_db = frequency_check(high)

# Duration of the file
duration = len(data) / sample_rate

# Create time array for x-axis
time = np.linspace(0., duration, len(data))
# Create a time array for x-axis
t = np.arange(len(data_in_db))
# t = np.arange(duration)
print(duration)
print(data)
print(len(data_in_db))
print(data_in_db)
print(len(t))
print(t)
print(time)
print(len(time))

# Plot the decibels over time
plt.figure(2)
plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')

# Mark the index of the maximum value
index_of_max = np.argmax(data_in_db)
value_of_max = data_in_db[index_of_max]
plt.plot(t[index_of_max], value_of_max, 'go', label='Max Value')

# Slice the array from the max value
sliced_array = data_in_db[index_of_max:]

#find a nearest value of less 5db
def find_nearest_value(array, value):
  array = np.asarray(array)
  idx = (np.abs(array - value)).argmin()
  return array[idx]

# Find the index and value of less 5dB
value_of_max_less_5 = value_of_max - 5
value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)
index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)[0][0]
plt.plot(t[index_of_max_less_5], value_of_max_less_5, 'yo', label='5dB Less')

# Find the index and value of less 25dB
value_of_max_less_25 = value_of_max - 25
value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)[0][0]
plt.plot(t[index_of_max_less_25], value_of_max_less_25, 'ro', label='25dB Less')

# Calculate RT20 and RT60
rt20 = t[index_of_max_less_5] - t[index_of_max_less_25]
rt60 = 3 * rt20


plt.xlabel('Time (s)')
plt.ylabel('Power (dB)')
plt.title('Decibels over Time')
plt.grid()
plt.legend()
plt.show()

print(f'The RT60 reverb time at freq {int(target_frequency)} Hz is {round(abs(rt60), 2)} seconds')

print(len(t))
print(len(data))


recording_conversion_testing.py
from scipy.io import wavfile
import scipy.io
import os
from pydub import AudioSegment

def convert_to_wav(input_file, output_file):
    audio = AudioSegment.from_file(input_file)

    # Ensure the output file has a .wav extension
    if not output_file.lower().endswith('.wav'):
        output_file = os.path.splitext(output_file)[0] + '.wav'

    # Export the audio as a WAV file
    audio.export(output_file, format='wav')
    print(f"Conversion successful. Saved as {output_file}")


script_directory = os.path.dirname(os.path.abspath(__file__))
wav_fname = os.path.join(script_directory, 'clap.mp3')
wav_fname2 = os.path.join(script_directory, 'clap_test.wav')

convert_to_wav(wav_fname, wav_fname2)

samplerate, data = wavfile.read(wav_fname2)

print(f"number of channels = {data.shape[len(data.shape) - 1]}")
print(f"sample rate = {samplerate}Hz")
length = data.shape[0] / samplerate
print(f"length = {length}s")


main.py
import os
import io
import random
import tkinter
from tkinter import *
from tkinter import filedialog
from itertools import count, cycle
import numpy as np
from PIL import Image, ImageTk
import matplotlib.pyplot as plt
import scipy
from pydub import AudioSegment

from model import Model
mdl = Model()

root = Tk()

# Update Window Size
def UpdateWindow():
    root.update()
    submitWidth = importButton.winfo_width() + fileNameLabel.winfo_width() + 4
    fileInfoWidth = fileInfoLabel.winfo_width() + 4
    photoWidth = imageLabel.winfo_width()
    windowWidth = max(submitWidth, fileInfoWidth, photoWidth, 300)
    root.geometry('%dx%d' % (windowWidth, layerHeight * 5 + imageLabel.winfo_height()))

# Upload Sound File
def UploadAction(event=None):
    global filename
    filedir = filedialog.askopenfilename()
    if 'filedir' in locals():
        if (filedir.lower().endswith('.wav') or filedir.lower().endswith('.mp3')):
            global wav, time
            wav = mdl.clean_audio_data(filedir)
            print('Selected:', filedir)
            filename = os.path.basename(filedir).split('/')[-1]
            fileNameLabel.config(text=filename)
            UpdatePicture(0)
            UpdateWindow()
            titleLabel.config(text="No Image Selected")
            # Import Images
            global img1, img2, img3, img4, img5, img6
            global amp1, amp2, amp3, amp4
            global rt60_1, rt60_2, rt60_3, rt60_4
            img1, rt60_1, amp1 = mdl.plotting(wav, 100, False, False)
            img2, rt60_2, amp2 = mdl.plotting(wav, 1000, False, False)
            img3, rt60_3, amp3 = mdl.plotting(wav, 7500, False, False)
            img4, rt60_4, amp4 = mdl.plotting(wav, 1000, False, True)
            img5, time = mdl.createWavForm(wav)
            img6, rt60_5, amp5 = mdl.plotting(wav, 1000, True, False)
            del filedir
            fileInfoLabel.config(text="Time: " + str(time) + " seconds; High Amp: N/A; RT60 Dif: N/A")
        else:
            fileNameLabel.config(text = "Invalid File Format")
            UpdatePicture(0)
            UpdateWindow()

# Update Checkbox Pictures
def UpdatePicture(x):
    if ((fileNameLabel.cget("text") != "No File Selected") and ((fileNameLabel.cget("text") != "Invalid File Format"))):
        match x:
            case 0:
                imageLabel.config(image="", text="Select A Data Visualization Option")
            case 1:
                imageLabel.config(image=img1, text="")
                titleLabel.config(text="Low Frequencies (100Hz) of " + filename)
                fileInfoLabel.config(text="Time: " + str(time) + " seconds; High Amp: " + str(amp1) + " db; RT60 Dif: " + str(rt60_1) + " seconds")
            case 2:
                imageLabel.config(image=img2, text="")
                titleLabel.config(text="Medium Frequencies (1000Hz) of " + filename)
                fileInfoLabel.config(text="Time: " + str(time) + " seconds; High Amp: " + str(amp2) + " db; RT60 Dif: " + str(rt60_2) + " seconds")
            case 3:
                imageLabel.config(image=img3, text="")
                titleLabel.config(text="High Frequencies (7500Hz) of " + filename)
                fileInfoLabel.config(text="Time: " + str(time) + " seconds; High Amp: " + str(amp3) + " db; RT60 Dif: " + str(rt60_3) + " seconds")
            case 4:
                imageLabel.config(image=img4, text="")
                titleLabel.config(text="Combined Frequencies of " + filename)
                fileInfoLabel.config(text="Time: " + str(time) + " seconds; High Amp: " + str(max(amp1,amp2,amp3)) + " db; (Average) RT60 Dif: " + str(round((rt60_1 + rt60_2+ rt60_3) / 3,2)) + " seconds")
            case 5:
                imageLabel.config(image=img5, text="")
                titleLabel.config(text="Waveform of " + filename)
                fileInfoLabel.config(text="Time: " + str(time) + " seconds; High Amp: " + str(max(amp1,amp2,amp3)) + " db; (Average) RT60 Dif: " + str(round((rt60_1 + rt60_2+ rt60_3) / 3,2)) + " seconds")
            case 6:
                imageLabel.config(image=img6, text="")
                titleLabel.config(text="Spectrogram of " + filename)
                fileInfoLabel.config(text="Time: " + str(time) + " seconds; High Amp: " + str(max(amp1,amp2,amp3)) + " db; (Average) RT60 Dif: " + str(round((rt60_1 + rt60_2+ rt60_3) / 3,2)) + " seconds")
            case _:
                imageLabel.config(image="", text="i literally have no idea how you got here")
    elif (fileNameLabel.cget("text") == "Invalid File Format"):
        imageLabel.config(image = "", text = "Invalid File Format")
    else:
        imageLabel.config(image = "", text = "No File Given")
    UpdateWindow()

# Create Import Button
importButton = Button(root, text='Open', command=UploadAction)
importButton.place(x=0, y=0)

# Show File Name If Imported
fileNameLabel = Label(root, text="No File Selected")
fileNameLabel.place(x=0, y=0)
root.update()

# Calculate window variables
layerHeight = importButton.winfo_height()
textHeight = (importButton.winfo_height() - fileNameLabel.winfo_height()) / 2

# Place File Name In Correct Position
fileNameLabel.place(x=importButton.winfo_width() + 2, y=textHeight)
fileInfoLabel = Label(root, text="Time: N/A; High Amp: N/A; RT60 Dif: N/A")
fileInfoLabel.place(x=2, y=layerHeight+textHeight)

# Graph Display
titleLabel = Label(root, text = "No Image Selected")
titleLabel.place(anchor=N, relx=.5, y=layerHeight*4+textHeight)

# Graph Display
imageLabel = Label(root, image = "", text = "No File Given")
imageLabel.place(x=0, y=layerHeight*5)

# Picture Options
lowrB = Button(root, text='Low Freq' , width = 10, command=lambda: UpdatePicture(1))
mediB = Button(root, text='Mid Freq' , width = 10, command=lambda: UpdatePicture(2))
highB = Button(root, text='High Freq', width = 10, command=lambda: UpdatePicture(3))
lowrB.place(anchor=N, relx=.15, y=layerHeight*2)
mediB.place(anchor=N, relx=.50, y=layerHeight*2)
highB.place(anchor=N, relx=.85, y=layerHeight*2)
totlB = Button(root, text='All Freq' , width = 10, command=lambda: UpdatePicture(4))
waveB = Button(root, text='Waveform' , width = 10, command=lambda: UpdatePicture(5))
ppppB = Button(root, text='Spectrogram', width = 10, command=lambda: UpdatePicture(6))
totlB.place(anchor=N, relx=.15, y=layerHeight*3)
waveB.place(anchor=N, relx=.50, y=layerHeight*3)
ppppB.place(anchor=N, relx=.85, y=layerHeight*3)

# Start Window
root.title("Audio Frequencies")
UpdateWindow()
root.resizable(False, False)
root.mainloop()

# Close Generated Images
if os.path.exists("converted_audio.wav"):
    os.remove("converted_audio.wav")
if os.path.exists("mono_audio.wav"):
    os.remove("mono_audio.wav")
if 'wav' in globals():
    if os.path.exists(wav):
        os.remove(wav)